dimod/higherorder/polynomial.py

```diff
@@ -16,11 +16,14 @@
 
 import itertools
 import collections.abc as abc
+from collections import defaultdict
 
 from numbers import Number
 
 import numpy as np
 
+import cytoolz as tl
```

Comment:
while `cytoolz` is a cool package, I don't think we can add it as a core ocean dependency. It seems supported-ish, but, for instance, [it only started releasing wheels in august/september](https://pypi.org/project/cytoolz/0.11.0/#files).

Reply:
Fair enough. They have been conda since 2016 though. Many science/ai/ml package are conda only/first.

I am also using these PRs to explore ideas even if they do get merged. Hope that is ok

Reply:
More than ok! It's much easier to have these conversations on a PR.

For conda, we actually have an issue about this https://github.com/dwavesystems/dwave-ocean-sdk/issues/144. I do think conda support in addition would be nice, but we also need to support pypi installs.

Reply:
Great! I would love to have dwave code in conda. Outside dwave I use conda almost exclusively. One adwantage is that the offical repos are curated so the quality tends to be higher than pypi. Also the conda (& now also mamba) package manager maintains global environment consistecy so one install does not break previously installed packages. Conda will (up/down) grade previouly installed packages if necessary. Conda is also language agnostic so you can distribute C, R, Julia, Rust packages with it. Main drawback is it can not install directly from git.

The reason I brought cond up though was more that cytoolz is more active/mature/maintained than it's pypi history would suggest. I believe dask was also long time in conda before it got released on pypi.

Reply:
Also conda gives you MKL compiled numerical libraries by default, which in turn means that native vectorised numpy/scipy operation run multithreaded by default (without GIL)

Reply:
The original pure python pytoolz package has a [PyPi release history](https://pypi.org/project/toolz/#history) going back to 2013.
It is a drop in replacement for cytoolz (or rather the other way around) and for this use case it is less than 2x slower than cytoolz so we still get >5x instead of >8x speedup and retains the nice code. At the moment this PR only uses tl.frequencies,
the other uses were replaced by generator expressions.

Reply:
It also looks like cytool zhas been in PiPy since 2014 https://pypi.org/project/cytoolz/#history****

Reply:
Yes, though they only released an sdist until september. Which means that users without python-dev installed would not be able to install Ocean.
We have to be extremely careful about introducing new dependencies, especially ones with compiled code. Because they need to work across all of our supported OS/python versions.

Reply:
Ok, I did not realize that the wheel availability was the issue. I thought the main concern was general maturity & support. If they are releasing wheels now what is the criterion for acceptance? Some period of time or some number of wheel releases?

Reply:
There isn't a hard algorithm. But experience over the last few years has taught us to be extremely cautious when introducing new dependencies to Ocean. Especially compiled dependencies. A significant fraction of our support requests come from dependency management. Another place this comes up, each time Python releases a new version we can only release a supporting version of Ocean once all of our dependencies also update. So for some packages like NumPy, they are so fundamental that we cannot possibly avoid using them. But otherwise we definitely err, pretty far, on keeping our dependency tree as small as possible.

Reply:
Fair enough. I do think that (cy)toolz is the kind universal nuts&bolts library linke numpy, scipy & pandas that is worth considering since it can be used everywhere and can reduces the need for resorting to C & cython